### ISSUE TITLE
Implement Garbage Collection Proposal for WebAssembly

### ISSUE THREAD

**Issue by fniephaus** [oracle/graal#9758]
### TL;DR

We plan to implement [the WebAssembly Garbage Collection proposal](https://github.com/WebAssembly/gc) in GraalWasm.

### Details

This proposal lets programs compiled to WebAssembly make use of the embedder's garbage collector. It achieves this by introducing types for structs and arrays and exposing new instructions that let WebAssembly modules allocate such objects.
These new objects will be managed by the runtime, outside of the module's linear memory. This means that when you compile a high-level language with GC to WebAssembly, you will no longer need to include a GC in the compiled module.
This, in turn, will lead to both smaller modules and faster execution, since you can rely on the mature GC used in GraalVM.

More details at https://github.com/WebAssembly/gc/blob/main/proposals/gc/Overview.md

**Comment by reneleonhardt** [comment]
With so many proposals like GC already implemented for WASM 2 by most parties, could you update the deprecated link?

https://github.com/WebAssembly/spec/blob/main/proposals/gc/Overview.md

https://github.com/WebAssembly/gc has been archived and WASM 3 has been completed.
https://webassembly.org/news/2025-09-17-wasm-3.0/

**Comment by jirkamarsik** [comment]
@reneleonhardt Thanks for pointing that out! The good news is that our GC implementation is now undergoing peer review and so should land in GraalWasm `master` soon. :-)

https://github.com/oracle/graal/pull/12701

**Comment by reneleonhardt** [comment]
Wow, amazing work, love that Graal is constantly improving it's WASM support, I believe it will be an important cornerstone for Java (for example deploying to Cloudflare Workers) ❤️

When you get in touch with webassembly.org, please let them sync the news and features overview to show what GraalWasm has already implemented.

https://webassembly.org/news/2025-09-17-wasm-3.0/
